File: python/rectangles/rectangles.py
"""Rectangles Parsing."""

import itertools


def rectangles(strings: list[str]) -> int:
    """Count all rectangles in a diagram."""
    # Map coordinates to chars.
    elements = {
        (x, y): char
        for y, line in enumerate(strings) for x, char in enumerate(line)
        if char in "+-|"
    }
    # Build a set of corners, horizontal and vertical pieces.
    corners = {point for point, char in elements.items() if char == "+"}
    horiz = {point for point, char in elements.items() if char == "-"} | corners
    vert = {point for point, char in elements.items() if char == "|"} | corners

    count = 0
    # For every corner pair, see if it forms a rectangle.
    for (x0, y0), (x1, y1) in itertools.product(corners, repeat=2):
        # Only count pairs with (x0, y0) < (x1, y1)
        if x0 >= x1 or y0 >= y1:
            continue
        # Check the top right, bottom left corners.
        if (x0, y1) not in corners or (x1, y0) not in corners:
            continue
        # Check the top and bottom sides.
        if any((x, y0) not in horiz for x in range(x0, x1 + 1)):
            continue
        if any((x, y1) not in horiz for x in range(x0, x1 + 1)):
            continue
        # Check the left and right.
        if any((x0, y) not in vert for y in range(y0, y1 + 1)):
            continue
        if any((x1, y) not in vert for y in range(y0, y1 + 1)):
            continue
        count += 1
    return count

File: go/rectangles/rectangles.go
package rectangles

type point struct{ x, y int }
type board struct {
	corners, horz, vert map[point]bool
}

func newBoard(diagram []string) board {
	corners := map[point]bool{}
	horz := map[point]bool{}
	vert := map[point]bool{}

	for y, line := range diagram {
		for x, char := range line {
			if char == ' ' {
				continue
			}
			p := point{x, y}
			switch char {
			case '+':
				corners[p] = true
				horz[p] = true
				vert[p] = true
			case '|':
				vert[p] = true
			case '-':
				horz[p] = true
			default:
				panic("Invalid char " + string(char))
			}
		}
	}
	return board{corners, horz, vert}
}

func (b board) isRect(topLeft, bottomRight point) bool {
	// topLeft must be above and left of bottomRight.
	if topLeft.x >= bottomRight.x || topLeft.y >= bottomRight.y {
		return false
	}
	// Check the other two corners exist.
	topRight := point{bottomRight.x, topLeft.y}
	bottomLeft := point{topLeft.x, bottomRight.y}
	if !b.corners[topRight] || !b.corners[bottomLeft] {
		return false
	}

	// Check the top and bottom edges exist.
	for x := topLeft.x; x <= topRight.x; x++ {
		if !b.horz[point{x, topLeft.y}] || !b.horz[point{x, bottomLeft.y}] {
			return false
		}
	}

	// Check the right and left edges exist.
	for y := topLeft.y; y <= bottomLeft.y; y++ {
		if !b.vert[point{topLeft.x, y}] || !b.vert[point{topRight.x, y}] {
			return false
		}
	}
	return true
}

// Count returns the number of rectangles in a diagram.
func Count(diagram []string) int {
	board := newBoard(diagram)

	count := 0
	// Iterate through all potential rectangle corners.
	for topLeft := range board.corners {
		for bottomRight := range board.corners {
			if board.isRect(topLeft, bottomRight) {
				count++
			}
		}
	}
	return count

	panic("Please implement the Count function")
}
